server/src/main/java/org/elasticsearch/action/support/CountDownActionListener.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.action.support;
+
+import org.elasticsearch.action.ActionListener;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Wraps another listener and adds a counter -- each invocation of this listener will decrement the counter, and when the counter has been
+ * exhausted the final invocation of this listener will delegate to the wrapped listener. Similar to {@link GroupedActionListener}, but for
+ * the cases where tracking individual results is not useful.
+ */
+public final class CountDownActionListener extends ActionListener.Delegating<Void, Void> {
+
+    private final AtomicInteger countDown;
+    private final AtomicReference<Exception> failure = new AtomicReference<>();
+
+    /**
+     * Creates a new listener
+     * @param groupSize the group size
+     * @param delegate the delegate listener
+     */
+    public CountDownActionListener(int groupSize, ActionListener<Void> delegate) {
+        super(Objects.requireNonNull(delegate));
+        if (groupSize <= 0) {
+            assert false : "illegal group size [" + groupSize + "]";
+            throw new IllegalArgumentException("groupSize must be greater than 0 but was " + groupSize);
+        }
+        countDown = new AtomicInteger(groupSize);
+    }
+
+    /**
+     * Creates a new listener
+     * @param groupSize the group size
+     * @param runnable the runnable
+     */
+    public CountDownActionListener(int groupSize, Runnable runnable) {
+        this(groupSize, ActionListener.wrap(Objects.requireNonNull(runnable)));
+    }
+
+    private boolean countDown() {
+        return countDown.updateAndGet(current -> {
+            if (current <= 0) {
+                throw new IllegalStateException("over-decrementing of count down, listener invoked too many times");
+            }
+            return current - 1;
+        }) == 0;
```

Comment:
I think this should assert it doesn't overshoot so that tests don't accidentally handle this exception. OTOH I wonder whether this should be an exception in prod code. How about something like this?

```suggestion
        final var result = countDown.getAndUpdate(current -> Math.max(0, current - 1));
        assert result > 0;
        return result == 1;
```

Reply:
++, great suggestion, added as a29aae9